### DISCUSSION TITLE
Source chain structure - atomic commit bundles

### DISCUSSION THREAD

**Discussion by mattyg** [holochain/holochain#2412] [General]
This one was brought up by @artbrock and I agree it makes a lot of sense and I feel it would reduce a lot of the pain in reasoning about and implementing validation code.

Currently all source chain commits within a single hdk_extern call are atomic -- they're either *all* committed or *none* are.

However, in the source chain itself, each action is a separate addition to the hash chain. This means that in validation, we need to validate each action individually, and aren't able to easily see the other actions it is intended to be committed with, and thus can't easily co-validate them.

For example you might commit an entry and a link to that entry in a single hdk_extern call. You then want to validate that both exist -- in your validation of the CreateLink action, you can do a `must_get_entry` to ensure the entry exists, but in the validation of CreateEntry action you have no way to validate that the link exists. 

So you basically have to validate the entry *first* and then validate the link second. But then what if someone committed just the entry and not the link to make it discoverable? You have a valid CreateEntry action on the dht that is effectively spam because its not discoverable in the way that those entries are expected to be.

The proposal is that we change the source chain structure from a chain of individual Actions to a chain of bundles of Actions, where every action in a bundle is being committed "simultaneously", and either all pass validation or none do.

Interestingly this gives us a capacity for simultaneous events in the source chain.

Anyway I'm very unclear about if / how exactly this would change validation, source chain querying and dht authority responsibility, and it seems like a very large scope change.

**Comment by pdaoust** [comment]
It just occurred to me as I was reading this that this effectively makes a source chain a blockchain! 😅  Anyhow...

Logically it makes sense. I've never been too worried about this because self-validation would catch the error at the link validation stage, back out the whole atomic commit, and fail to publish the entry. But you're right, that doesn't cover malicious data, which is half the point of validation.

It would require rethinking of something. The DHT needs to balance lookup efficiency (DHT addresses are meaningful enough to be used as graph primitives) with security (DHT addresses are random enough to thwart collusion) and load balancing (not too much data goes to any one peer).

All this is to say that the validation callback would have to be rethought. I can see a couple options:

* go back to validating actions like in the Redux days, rather than validating ops. Or rather, validate bundles of actions.
* validate ops as now, but pass in the rest of the commit bundle data as explicit dependencies.

Either way, meaningful validation would require sending more data to more authorities -- for a `CreateEntry` and a `CreateLink`, five ops get sent to ten authorities for a total of fifty ops to be processed by the DHT (assuming a redundancy factor of only 1). In comparison, what we have now is five ops, each of which gets sent to only one authority, for a total of five ops to be processed.

Granted, many of those five ops have redundant data (`RegisterAgentActivity`, `StoreEntry`, and `StoreRecord` all have essentially the same data) and could be collapsed for validation... which is starting to sound a lot like just validating the action itself rather than the op.

Validating actions would take away the ability to allow each authority to focus on what it knows best (e.g., agent activity validates the `weight` field for rate limiting). One thought is that not all data has to be sent to authorities. We already see this with `EntryDef.cache_at_agent_activity`; a lot of fields could be left blank for this or that authority using similar flags like `EntryDef.cache_at_record`. In validation, you skip validating a certain bit if it's `None`, which is an indication that it shouldn't matter to you as an authority. This might actually be a nice way of abstracting ops away while still permitting authorities to validate with granular focus on their domains of authority -- I've found op-based validation to be kinda unwieldy.

Additionally, a zome function could opt into/out of atomic bundle validation with some sort of `atomic_validation` flag on `create`/`update`/`delete` similar to `relaxed_chain_top_ordering` -- if any of the writes has that flag, all of them are treated as atomic for the purpose of validation.

Anyhow, not saying any of this is correct, just brainstorming and trying to identify areas of concern.

**Reply by maackle** [reply]
Yes, it's a lot to think about! Some thoughts after reading this:

> validate ops as now, but pass in the rest of the commit bundle data as explicit dependencies.

This is the route I imagine, but it would certainly require huge changes.

> Granted, many of those five ops have redundant data

Yes, the way we're doing things now is a bit inefficient. Really, every op should just have references to the action and entry (what is currently `DhtOpLite`), with action and entry sent separately to avoid redundancy when sending several ops to the same node.

If we consider the subject matter of what ops are "about" to change from actions and entries, to bundles of actions and entries, then perhaps the ops would reference the actions/entries in a hierarchical fashion, something like `{bundle_hash}/{component_index}`, so that the op authorities could store only the subset of the bundle necessary for validation and performing their authority duties, but the entire bundle could still be reconstructed in its original form with knowledge of how many components it has. But this requires so many changes, like allowing an extra N bytes for the component index, changing the meaning of DHT location to include that index, and more..
